cirq-core/cirq/transformers/transformer_api_test.py

```diff
@@ -0,0 +1,275 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from unittest import mock
+
+import cirq
+from cirq.circuits.circuit import CIRCUIT_TYPE
+from cirq.transformers import LogLevel
+
+import pytest
+
+
+@cirq.transformer
+class MockTransformerClassCircuit:
+    def __init__(self):
+        self.mock = mock.Mock()
+
+    def __call__(self, circuit: cirq.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
+        self.mock(circuit, context)
+        return circuit
+
+
+@cirq.transformer
+class MockTransformerClassFrozenCircuit:
+    def __init__(self):
+        self.mock = mock.Mock()
+
+    def __call__(
+        self, circuit: cirq.FrozenCircuit, context: cirq.TransformerContext
+    ) -> cirq.FrozenCircuit:
+        self.mock(circuit, context)
+        return circuit
+
+
+@cirq.transformer
+class MockTransformerClassGeneric:
+    def __init__(self):
+        self.mock = mock.Mock()
+
+    def __call__(self, circuit: CIRCUIT_TYPE, context: cirq.TransformerContext) -> CIRCUIT_TYPE:
+        self.mock(circuit, context)
+        return circuit
+
+
+@cirq.transformer
+def mock_transformer_method_circuit(
+    circuit: cirq.Circuit, context: cirq.TransformerContext
+) -> cirq.Circuit:
+    if not hasattr(mock_transformer_method_circuit, 'mock'):
+        mock_transformer_method_circuit.mock = mock.Mock()  # type: ignore
+    mock_transformer_method_circuit.mock(circuit, context)  # type: ignore
+    return circuit
```

Comment:
I'd suggest having a factory to create these transformers instead, so the mock is not part of a static object:
```python
def make_circuit_transformer_func() -> TRANSFORMER['cirq.Circuit']:
    my_mock = mock.Mock()
    @cirq.transformer
    def func(circuit: cirq.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
        my_mock(circuit, context)
        return circuit
    func.mock = my_mock
    return func
...
@pytest.mark.parametrize('transformer', [make_circuit_transformer_func(), ...])
def test...
```
This will avoid bugs if we add more tests that use circuit transformer funcs. Same thing for frozen circuit transformer funcs below.